botocore/args.py

```diff
@@ -472,6 +491,66 @@ def _ensure_boolean(self, val):
         else:
             return val.lower() == 'true'
 
+    def _build_endpoint_resolver(
+        self,
+        endpoints_ruleset_data,
+        partition_data,
+        client_config,
+        service_model,
+        endpoint_region_name,
+        region_name,
+        endpoint_url,
+        endpoint,
+        is_secure,
+        endpoint_bridge,
+        event_emitter,
+    ):
+        if endpoints_ruleset_data is None:
+            return None
+
+        # The legacy EndpointResolver is global to the session, but
+        # EndpointRulesetResolver is service-specific. Builtins for
+        # EndpointRulesetResolver must not be derived from the legacy
+        # endpoint resolver's output, including final_args, s3_config,
+        # etc.
+        s3_config_raw = self.compute_s3_config(client_config) or {}
+        service_name_raw = service_model.endpoint_prefix
```

Comment:
This mirrors what the existing code was doing at this line: https://github.com/boto/botocore/blob/730f1ebe4b209e676949263af4a6a6865b6cb0ce/botocore/args.py#L167

Due to refactoring for this PR this is no longer obvious: I used the `_raw` suffix in a few variable names to indicate that this is the raw/original/unmodified version of a local variable in the `ClientArgsCreator.get_client_args` function, before any modifications applied by the legacy endpoint resolver. The line you are commenting on is the complement to the one I linked above.

That said, I agree with your concern. The reason this has been working and will continue working is that the service name determined in this slightly wrong way is only used to compare against `s3`, `s3-control`, and `sts`. In fact, changing it would break the logic because [this line](https://github.com/boto/botocore/blob/730f1ebe4b209e676949263af4a6a6865b6cb0ce/botocore/client.py#L544) references the host prefix `s3-control` and _not_ the folder name `s3control`.